### ISSUE TITLE
Fix CMake Build for MSVC 15 2017

### ISSUE THREAD

**Issue by LaylDupont** [python-cmake-buildsystem/python-cmake-buildsystem#233]
Build for MSVC 15 2017 does not work right out of the box and could be a little more documented.

CMake project configuration and generation is going fine.

`cmake ../python-cmake-buildsystem`

It generates project files for Visual Studio and Python.sln Solution.
It is worth documenting what one should do at this point to build the solution (i assumed i had to run this:)

`MSBuild.exe python.sln`

However i get an error when i let default configuration:

> error MSB4126

So i tried to run with explicit configuration:

`MSBuild.exe python.sln /p:Configuration=Release /p:Platform=x86`

However i get the same error:

> error MSB4126

Then i tried the following tip i saw somewhere on SO concerning the same error:

```
SET "PLATFORM="
MSBuild.exe python.sln
```

However i get the following error:

> winnt.h(154): fatal error C1189: #error:  "No Target Architecture"

So i tried to include this patch:
in file _iomodule.c from Python 3.6.4 sources as pulled by default

From:
> l.23: #ifdef MS_WINDOWS
> l.24: #include <consoleapi.h>
> l.25: #endif

To:
> l.23: #ifdef MS_WINDOWS
> l.24: //PATCH:START
> l.25: #include <windows.h>
> l.26: //PATCH:END
> l.27: #include <consoleapi.h>
> l.28: #endif

When doing so i get "build success" with output python.exe, python36.lib, pyconfig.h, so now i can get to chose which modules i want and get that static python for embedding
By the way i do not get any of these errors with the regular python build.

In short:
Is it possible to improve documentation for build on windows ?
Is it possible to add a patch for this ?

**Comment by jcfr** [comment]
Thanks for reporting the issue :+1: 

Here is a PR adding testing for VS 2017. See https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/234

> Is it possible to improve documentation for build on windows ?

Definitively

What about updating the "Build" section with the following:

```
# Unix
cd ~/scratch
mkdir -p python-build && mkdir -p python-install
cd python-build
cmake -DCMAKE_INSTALL_PREFIX:PATH=${HOME}/scratch/python-install ../python-cmake-buildsystem
make -j10
make install

# Windows
cd %HOME%/scratch
mkdir python-build
mkdir python-install
cd python-build
cmake -G "Visual Studio 15 2017 Win64" -DCMAKE_INSTALL_PREFIX:PATH=%HOME%/scratch/python-install ../python-cmake-buildsystem
cmake --build . --config Release -- /m
cmake --build . --config Release --target INSTALL
```

> Build error 

After, enabling testing for `Visual Studio 2017`

The build are successful: https://ci.appveyor.com/project/jcfr/python-cmake-buildsystem/build/1.0.348

I suggest you follow the approach suggested above

**Comment by LaylDupont** [comment]
Thank you for your quick answers, indeed these build instructions for Windows seem to be what should work however i stumble on the same error as before during build, when running:

`cmake --build . --config Release -- /m`

I get:

> winnt.h(154): fatal error C1189: #error: "No Target Architecture"

Which apparently has something to do with how windows API headers are included in python source code and can be solved with the patch i proposed for 3.6 in the first post for this issue.



Another issue i ran into before and that made me use MSBuild directly instead of cmake --build was that if i ran:

`cmake --build .`

instead of

'cmake --build . --config Release -- /m'

I would then get this error:

> error MSB8020: Les outils de génération pour Visual Studio 2010 (ensemble d'outils de plateforme ='v100') sont introuvables.

in short cmake targets build tools for MSVC 2010 tools instead of MSVC 2017 even though i used the Generator "Visual Studio 15 2017 Win64".
I've had this issue on two different clean setups on Windows where only MSVC 2017 tools are installed, Maybe this command is wrong, or this might be an issue from CMake itself, but i do not experience this error with cmake builds for other libraries.

If the automated builds work, then a patch is illegitimate. In that case i am clueless about how to get the CMake Build System with MSVC 15 2017 on my configuration.

**Comment by LaylDupont** [comment]
Ok so after spending too much time on that, here are more news ;) :

`cmake --build .`

targets debug build by default but the generator only generated a release config somehow so MSBUILD is lost and outputs obscure errors ( those related to MSVC 2010 )
I guess this can be overcome by setting CMAKE_BUILD_TYPE at generation / configuration time, that was a noob error here, although this behaviour is not standard for cmake library builds.

An interesting point is It seems that the error due to the include of <windows.h> in iomodule.c was solved in 3.7.0.
Actually my patch is exactly the same as the changes they made to that file.
So this is a change that is already taken into account in cpython that shouldn't be patched in the future here.
This change was made to solve problems due to windows SDK updates so it seems like your VMs from where you run your tests are a little outdated.

So i thought i could use your branch "support-build-37-and-38" since it would typically solve this, but it doesn't seem to be ready to work either ! + it has troubles with some external libs because it is looking for files that were removed from Python/Modules/zlib

To summarize:
The python cmake buildsystem is able to work under Windows and for Visual Studio 2017 although a patch is possible for 3.6 ( at least ), it corresponds to exact same changes that were already made in 3.7.0.

For my needs i will be using my patch.
Should i make another issue or push the patch for this specific system / generator / compiler ?

Thanks ;)

**Comment by jcfr** [comment]
Thanks for following up.

### Few comments

* When using Visual Studio, ` CMAKE_BUILD_TYPE` should not be set at generation. Indeed, Visual Studio is a multi-configuration IDE and the config should be selected at build time. I will add a note to the README.

* Debug build are not supported (yet) on windows. See #12

> Which apparently has something to do with how windows API headers are included in python source code and can be solved with the patch i proposed for 3.6

Could you for the `python-cmake-buildsystem/cpython` project and push a commit one top of this branch: https://github.com/python-cmake-buildsystem/cpython/tree/3.6.2-patches

This would greatly help moving this forward.


> Actually my patch is exactly the same as the changes they made to that file.

It would be great to reference here the commit from the upstream python github repository.


### Test build

Could you try to build locally with the option `BUILD_LIBPYTHON_SHARED:BOOL=ON` ? This would allow to build like it was done on appveyor: https://ci.appveyor.com/project/jcfr/python-cmake-buildsystem/build/1.0.348

I would like to confirm that a build is successful for you when using the same settings.

**Comment by LaylDupont** [comment]
I submitted a Pull-Request with the patch for Fixing Windows build of Python, reverberating changes that were made upstream for 3.7 here in the 3.6.2 patches in order to allow build using the latest WinSDK.

Here is the reference of the commit upstream, also mentionned in the commit & pull request i submitted.
[https://github.com/python/cpython/commit/c0ee341b29bd7d978b49272a2c0e2dcfa77404d5](https://github.com/python/cpython/commit/c0ee341b29bd7d978b49272a2c0e2dcfa77404d5)

### Test build

Unfortunately the build was not successful using these settings, due to the same error as before.



Merci d'avoir pris le temps de répondre, et merci pour ton travail sur le cmake buildsystem, ça simplifie beaucoup les choses pour builder python en static pour l'embarquer ! ;)

**Comment by jcfr** [comment]
Build issue with new install of VS2017 has been fixed in https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/commit/2cd115b830e01305a0cd359e468630d1c6f2c9ed

**Comment by jcfr** [comment]
To summarize:

> Is it possible to improve documentation for build on windows ?

This was done in https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/commit/75b8384d73575eed2a46ac3e75c6690502380d04

> Is it possible to add a patch for this ?

This was done in https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/commit/55c8825d9454d18796e409d2cb6df60fcee12aaf

Thanks again for your patience.
